# Multipartitions #

This notebook computes values of $p(a, b)$, that is, the number of partitions of Gaussian integers $a + bi$ into parts $u + vi$ so that $u > 0$ and $v > 0$.

## References ##
* M. S. Cheema and T. S. Motzkin, "Multipartitions and multipermutations," Proc. Symp. Pure Math. 19 (1971), 39-70, eq. (3.1.3).
* The Online Encyclopedia of Integer Sequences (OEIS) entry for p(a, b): [A090806](http://oeis.org/A090806)
* Main diagonal: OEIS [A108469](http://oeis.org/A108469)

In [1]:
import pandas as pd

from multip import timed, memoize
from fractions import gcd

## Recurrence ##

This is equation (3.1.3) in Cheema and Motzkin which defines the recurrence used in this notebook. For the purposes of this notebook, $s = 2, U(n_1, n_2) = p(n_1, n_2)$. The outer sum is over $k_1$ and $k_2$ and the inner sum is over $t$ such that $t$ divides $\text{gcd}(k_1, k_2)$.

$\displaystyle (3.1.3)\hspace{2pc}n_1U(n_1,\ldots,n_s)=\sum_{k_i>0}\big(U(n_1-k_1,\ldots,n_s-k_s)\sum_t k_1/t\big);$

In [2]:
@memoize
def gcd_(a, b):
    return gcd(a, b)

@memoize
def inner(k1, k2):
    D = gcd(k1, k2)
    return sum([k1//t for t in range(1, gcd_(k1,k2)+1) if D//t == D/t])

@memoize
def p(a, b):
    if b > a:
        return p(b, a)
    if(a==0 and b==0):
        return 1
    if(b==0):
        return 0
    if(a==1 or b==1):
        return 1
    if(b==2):
        return a//2 + 1 # Exercise: prove this!
    return sum([p(a-k1, b-k2)*inner(k1, k2) for k1 in range(1,a+1) for k2 in range(1,b+1)])//a

In [3]:
NN = 101
data = timed(lambda N:[[p(x, y) for y in range(1,N)] for x in range(1,N)])(NN)
diag = [data[i][i] for i in range(NN-1)]
pd.DataFrame(data, columns=range(1,NN), index=range(1,NN))

0:00:17.097694


1    2    3     4      5       6        7        8         9         10   \
1      1    1    1     1      1       1        1        1         1         1   
2      1    2    2     3      3       4        4        5         5         6   
3      1    2    4     5      7       9       11       13        16        18   
4      1    3    5     9     12      17       22       29        35        44   
5      1    3    7    12     20      28       40       53        70        89   
6      1    4    9    17     28      45       64       91       123       164   
7      1    4   11    22     40      64      100      144       204       278   
8      1    5   13    29     53      91      144      220       317       449   
9      1    5   16    35     70     123      204      317       478       688   
10     1    6   18    44     89     164      278      449       688      1025   
11     1    6   21    52    112     211      373      614       971      1471   
12     1    7   24    63    137     271      486      827      1332      2069   
13     1    7   27    73    168     337      628     1087      1797      2840   
14     1    8   30    86    201     419      794     1413      2375      3837   
15     1    8   34    98    240     510      995     1802      3101      5087   
16     1    9   37   114    282     619     1229     2280      3983      6664   
17     1    9   41   128    331     739     1507     2841      5067      8602   
18     1   10   45   146    383     882     1825     3516      6366     10995   
19     1   10   49   163    443    1036     2199     4301      7930     13890   
20     1   11   53   184    507    1217     2622     5231      9780     17404   
21     1   11   58   203    579    1414     3112     6299     11980     21595   
22     1   12   62   227    656    1640     3664     7550     14549     26615   
23     1   12   67   249    742    1884     4295     8975     17566     32533   
24     1   13   72   276    833    2164     5000    10624     21060     39533   
25     1   13   77   301    935    2463     5801    12489     25117     47710   
26     1   14   82   331   1042    2803     6690    14628     29776     57281   
27     1   14   88   359   1160    3167     7690    17030     35142     68361   
28     1   15   93   393   1285    3575     8796    19764     41254     81220   
29     1   15   99   424   1422    4011    10031    22816     48239     95993   
30     1   16  105   461   1566    4499    11389    26263     56149    113000   
..   ...  ...  ...   ...    ...     ...      ...      ...       ...       ...   
71     1   36  491  3917  22400  101488   387562  1296704   3905416  10792075   
72     1   37  504  4068  23451  107176   412264  1388966   4209532  11701171   
73     1   37  517  4213  24544  113061   438260  1486482   4533612  12675060   
74     1   38  530  4371  25668  119244   465539  1589789   4878527  13718392   
75     1   38  544  4523  26835  125642   494211  1698851   5245595  14834467   
76     1   39  557  4689  28035  132351   524272  1814246   5635739  16028399   
77     1   39  571  4848  29280  139289   555832  1935934   6050398  17303816   
78     1   40  585  5021  30559  146561   588886  2064528   6490594  18666305   
79     1   40  599  5188  31885  154071   623556  2199997   6957847  20119883   
80     1   41  613  5369  33247  161934   659832  2342988   7453287  21670632   
81     1   41  628  5543  34657  170056   697841  2493466   7978570  23322966   
82     1   42  642  5732  36105  178545   737580  2652127   8534888  25083536   
83     1   42  657  5914  37603  187308   779175  2818936   9124039  26957168   
84     1   43  672  6111  39140  196464   822624  2994625   9747338  28951095   
85     1   43  687  6301  40730  205905   868065  3179167  10406681  31070634   
86     1   44  702  6506  42360  215760   915491  3373336  11103509  33323620   
87     1   44  718  6704  44044  225922   965045  3577104  11839876  35715866   
88     1   45  733  6918  45771  236

In [4]:
diag

[1,
 2,
 4,
 9,
 20,
 45,
 100,
 220,
 478,
 1025,
 2169,
 4534,
 9364,
 19125,
 38646,
 77306,
 153173,
 300765,
 585518,
 1130612,
 2166284,
 4120062,
 7780817,
 14595364,
 27201794,
 50383690,
 92768457,
 169835952,
 309223286,
 560036477,
 1009124256,
 1809401232,
 3228940199,
 5735730725,
 10143456171,
 17861290223,
 31320395315,
 54699578899,
 95155744401,
 164903402689,
 284717811826,
 489817865949,
 839715146428,
 1434653959110,
 2442973258894,
 4146517217485,
 7015790971640,
 11834008865484,
 19901300509887,
 33370001915211,
 55793785786546,
 93024785473819,
 154675690989307,
 256496960062127,
 424232925096922,
 699862593674492,
 1151677339554945,
 1890518818157774,
 3095886527678785,
 5057825538844753,
 8243993501352917,
 13406824700552985,
 21754422747234735,
 35222547132631962,
 56906665201829593,
 91746972850520880,
 147612482736006935,
 237013531620962926,
 379802132549946504,
 607422488181757786,
 969589935603493072,
 1544767826010979783,
 2456574384729305274,
 389943344

The above calculation with `NN = 101` took about 20 seconds in a virtual machine on a Mid 2013 Macbook Air using a Python 3 kernel. A table with `NN = 301` can be generated in about the same time on the same hardware using [pypy](http://pypy.org/). See also the accompanying `multip` module.